API for tool management using npm

heralden opened the pull request:
This PR adds REST API endpoints for managing installation/updating of tools by executing npm commands through the shell.

Since we currently don't have authentication for this, we probably shouldn't merge this until this is in place.

If you wish to test it, make sure to add `:enable-unsafe-routes true` to your [environment](https://github.com/yogthos/config), as otherwise the unsafe routes for this API won't be added to the server.

heralden committed:
Experimental API wrapping tool management with npm

heralden committed:
Respond with new tools list to /api/tools/* requests

This saves the client from having to make another request to get the new
tools list.

heralden committed:
Delete tools package.json and initialise it automatically

heralden committed:
Use auth to only allow superusers to use npm tool API

heralden committed:
Message from tool API for older intermines without superuser flag

heralden committed:
Fix /api/tools/path endpoint

heralden committed:
Only initially install tool npm packages under intermine scope

This is to prevent creation of maliciously installed packages with the
bluegenes-intermine-tool keyword. If we filter down to the intermine
scope, only we have the ability to add to this.

heralden commented:
All features are in place now. Should be tested with https://github.com/intermine/bluegenes/pull/383.
Using `:enable-unsafe-routes true` is no longer necessary. All you have to do is `lein install` this version of bluegenes-tool-store to use it with bluegenes.

yochannah commented on src/clj/bluegenes_tool_store/auth.clj:
```diff
@@ -0,0 +1,51 @@
+            [config.core :refer [env]]
+            [clojure.string :as string]))
+
+(defn check-priv
```
This function is heavily commented - which is awesome - but docstrings make it easier to generate automated documentation, e.g. for codox - https://github.com/weavejester/codox. A short sentence describing what the function does would be sufficient.

yochannah commented on src/clj/bluegenes_tool_store/auth.clj:
```diff
@@ -0,0 +1,51 @@
+          ;; The InterMine instance is too old to support returning the
+          ;; superuser flag. We will instead give the user a useful message.
+          (response/not-implemented
+            {:error (str "Managing BlueGenes tools automatically is only supported on InterMine version 5.0.0 or newer. You can still perform the tool operation manually by running '"
```
perfect

yochannah commented on src/clj/bluegenes_tool_store/package.clj:
```diff
@@ -0,0 +1,33 @@
+            [taoensso.timbre :refer [infof]]))
+
+(defn get-all-npm-tools
+  "Send an HTTP request to the NPMJS API to get a list of bluegenes tools."
```
What happens if npm's api service isn't available? We could theoretically try falling back on another cdn if there is one (or does, say, yarn just use npm anyway?) but a simple error message is probaby sufficient.

yochannah commented on src/clj/bluegenes_tool_store/package.clj:
```diff
@@ -0,0 +1,33 @@
+  "Call NPM through the shell to install all bluegenes tools from the NPM registry."
+  []
+  (let [packages (mapv (comp :name :package) (get-all-npm-tools))]
+    (tools/sync-sh-req (into ["npm" "install" "--save"] packages)
```
given how very good npm is at breaking between versions, it might be nice to have some error handling here too :) I'd include the npm and node versions in any error messages if possible. Another possible path is that npm isn't installed or is a very old version.

yochannah commented on src/clj/bluegenes_tool_store/package.clj:
```diff
@@ -0,0 +1,33 @@
+  []
+  (let [config-path (tools/tools-path :config :file true)]
+    (when-not (.exists config-path)
+      (infof "Tool config %s does not exist. It will automatically be created and all bluegenes tools will be installed." config-path)
```
I couldn't quite work out from reading the code what the default path was going to be if there wasn't a tools path variable set. Is it dynamic, or could we add it in a comment?

heralden replied:
The bluegenes tool path? The default is specified in *config/defaults/config.edn*.

yochannah replied:
a comment above this info output would be a good place to specify the default and how to change it...

yochannah commented on src/clj/bluegenes_tool_store/tools.clj:
```diff
@@ -46,33 +54,93 @@
              :npm (get-in package [:name])}
      :config config
-     :package package
+     :package (select-keys package [:description :license :homepage :name :author :version])
```
good call, there was a lot of un-needed stuff in the entire package json ;)

yochannah commented on src/clj/bluegenes_tool_store/tools.clj:
```diff
@@ -46,33 +54,93 @@
+          (.unlock lock)))
+      (response/service-unavailable {:error "A tool operation is already in progress. Please try again later."}))))
+
+(defn install-package
```
I suspect we might need more error handling around the install/update functions in this file, too.

yochannah reviewed (changes requested):
first look through - this looks really well-thought-out for the "happy path" but I think we need a bit more error handling around scenarios like old npm versions and no npm installed at all. Knowing the way people work - this is a highly likely scenario, so at the very least we should add clear error messages for "please install npm and try again" and/or "the package failed to install/upgrade, here's the error messages and you used npm version x.y to install it."

heralden committed:
Remove dependency on npm for managing tools and make things robust

heralden committed:
Update gitignore

heralden committed:
Docstring and error message

heralden commented:
@yochannah check latest commits to see if the changes meet your requests.

heralden committed:
Fix deleting packages and improve errors

heralden committed:
Default to ./tools path and save tool versions as string

heralden committed:
Simple lein tools CLI for tool management

yochannah commented on src/clj/bluegenes_tool_store/tools.clj:
```diff
@@ -1,78 +1,206 @@
+    "uninstall" (package-operation :uninstall args)
+    (println "Install BlueGenes tools from the command line.
+
+USAGE: lein tools [ARGS...]
```
This is gorgeous! It should be in the readme, as a link to this comment if nothing else.

heralden replied:
Just a note: the `lein tools` commands won't give you good feedback as the error handling was made with the tool store in mind (ie. installing a nonexistant package will give you a "network error" and installing a package succesfully will just silently complete). I thought this was alright since the commands would just be used temporarily by us to manage *http://bluegenes.apps.intermine.org/* until flymine gets updated.
Do you still think we should mention it in the readme?

yochannah reviewed (approved):
- Tested lein prod:
  - http://localhost:5001/api/tools/all returned the tools list as expected
  - spot-checked a preview image url returned by the json tools/all list above, it worked as expected. 
- Tiny tweak request: I can see that the tools automatically downloaded into my tools folder. Nice. We should add this to the readme
- The comments are beautiful.

yochannah commented:
I think we should, but include the explanation you've just included about
the intent. Otherwise if you or I don't remember things properly, we'll
have to go code spelunking to find it, which isn't the best use of time :)

On Tue, 12 Nov 2019 at 12:30, Herald <notifications@github.com> wrote:

> *@uosl* commented on this pull request.
> ------------------------------
>
> In src/clj/bluegenes_tool_store/tools.clj
> <https://github.com/intermine/bluegenes-tool-store/pull/1#discussion_r345177875>
> :
>
> > +;;;;;;;;;;;;;;;
> +;; CLI Usage ;;
> +;;;;;;;;;;;;;;;
> +
> +(defn -main
> +  "This main function is for managing tools from the command line via the
> +  leiningen `lein tools` alias."
> +  [cmd & args]
> +  (timbre/set-level! :info)
> +  (case cmd
> +    "init"      (force-initialise-tools)
> +    "install"   (package-operation :install args)
> +    "uninstall" (package-operation :uninstall args)
> +    (println "Install BlueGenes tools from the command line.
> +
> +USAGE: lein tools [ARGS...]
>
> Just a note: the lein tools commands won't give you good feedback as the
> error handling was made with the tool store in mind (ie. installing a
> nonexistant package will give you a "network error" and installing a
> package succesfully will just silently complete). I thought this was
> alright since the commands would just be used temporarily by us to manage *http://bluegenes.apps.intermine.org/
> <http://bluegenes.apps.intermine.org/>* until flymine gets updated.
> Do you still think we should mention it in the readme?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/intermine/bluegenes-tool-store/pull/1?email_source=notifications&email_token=ACGXRDWD7WACLAKXYLC7GR3QTKOWRA5CNFSM4IJMOM62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCLHB4SY#discussion_r345177875>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGXRDRGOBZC2XTPTLJFBLTQTKOWRANCNFSM4IJMOM6Q>
> .
>

heralden committed:
Update README with notes on initialisation and CLI

We remove nodejs and npm as it's no longer required for
bluegenes-tool-store. (It is only required for bluegenes.)

heralden commented:
I updated the README.
